trawler/src/main/java/io/m9sweeper/trawler/ScanRunner.java

```diff
@@ -183,21 +195,30 @@ public void scan() throws Exception {
                     .policyId(imageTrawlerResultDtos.get(0).getPolicyId());
             List<ImageTrawlerResultDto> outdatedImageIdArray = new ArrayList<ImageTrawlerResultDto>();
             outdatedImageIdArray.add(outdatedImageIdResult);
-            saveScanResults(outdatedImageIdArray);
-        } else {
-            saveScanResults(imageTrawlerResultDtos);
+            compliant &= saveScanResults(outdatedImageIdArray);
+        }
+
+        // If we didn't encounter an error, save the results
+        // Note that if the image hash changed and we didn't encounter an error this will result in sending 2 results,
+        // The first being the error for the old hash, and this one which will save the actual scan under
+        // the correct hash.
+        if (!errorEncountered) {
+            compliant &= saveScanResults(imageTrawlerResultDtos);
+        }
+
+        if(!compliant) {
+            throw new NoncompliantException("Image Is not compliant", imageTrawlerResultDtos);
         }
     }
 
-    private void saveScanResults(List<ImageTrawlerResultDto> imageTrawlerResultDtos) throws Exception {
+    /** Returns whether or not the saved scan was compliant or not */
+    private boolean saveScanResults(List<ImageTrawlerResultDto> imageTrawlerResultDtos) throws Exception {
         TrawlerScanResults body = new TrawlerScanResults();
         body.setData(imageTrawlerResultDtos);
         if (TrawlerConfiguration.getInstance().getDebug()) {
             System.out.println("Saving Scan Results: " + body.toString());
         }
         ImageScanResultSaveResponse complianceResponse = apiInstance.imageControllerSaveImageScanResults(body, new BigDecimal(message.getCluster().getId()), new BigDecimal(message.getImage().getId()));
-        if (complianceResponse.getData() != null && !complianceResponse.getData().isComplaint()) {
-            throw new NoncompliantException("Image Is not compliant", imageTrawlerResultDtos);
-        }
+        return complianceResponse.getData() == null || complianceResponse.getData().isComplaint();
```

Comment:
Now made this return a boolean, and use the boolean in the main body to throw this error. This prevents the error from short-circuiting and not returning both scan results when needed.